### PR TITLE
Add a taxonomy for hiding lessons

### LINKED ISSUE #2735: Hidden Lessons appear in search results
This production Lesson is set to be hidden:

Image

We added a new post meta `_lesson_archive_excluded` to the Lesson post type, to enable to hiding specific Lessons (eg. quizzes) from archive and search views.

This works in local development and for the lessons archive in production, **but not with Jetpack search running in production**. I have tested this on the [staging site](https://learn.wordpress.org/test/) without Jetpack Search running and it works.

### Steps to reproduce

1. Log in to production
2. Create a Lesson and privately publish it. This should mean admins can see it but not regular users. Use a name you can easily search for later.
4. On the frontend navigate to the [Lessons archive](https://learn.wordpress.org/lessons/?new-theme=1) and locate your Lesson in the grid.
5. In the admin check the post meta box in the sidebar for 'Hidden Lesson -> Exclude this lesson from the archive'
Image
6. Refresh the frontend archive and the Lesson should be hidden
7. Search for your Lesson using the search box

#### Expected
The lesson remains hidden

#### Actual
The lesson is included in the search results

### PR DESCRIPTION
**Pull request by adamwoodnz**
Attempt to fix #2735 

Using post meta to hide lessons in search results doesn't work with Jetpack classic search, so this PR attempts to fix this by using a taxonomy, which are indexed. As suggested by @gibrown [here](https://github.com/WordPress/Learn/issues/2577#issuecomment-2248905487) I've used an [existing Jetpack taxonomy 'show'](https://github.com/Automattic/jetpack/blob/36b2d5232e2ec3a1ad14034ab673fddce3acab97/projects/packages/sync/src/modules/class-search.php#L1479) so that we don't need to add a new taxonomy.

I've left the existing post meta field for now while we migrate data to the new taxonomy, but a deprecation label has been added.

Screenshot 2024-08-08 at 2 47 43 PM

### Testing

This can only be tested in sandbox, and I think it will need to be shipped to be fully tested, due to the [Jetpack elastic search sync](https://jptools.wordpress.com/debug/?url=https://learn.wordpress.org/test/#async_get_sync_validator) needing to be run. I've tried syncing before shipping and it doesn't seem to work.

### CODE REVIEW

File: wp-content/plugins/wporg-learn/js/lesson-archive-excluded-meta/index.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ const LessonArchiveExcludedMeta = () => {
 	const [ lessonExcluded, setLessonFeatured ] = useState( postMetaData?._lesson_archive_excluded === EXCLUDED );
 
 	return (
-		<PluginDocumentSettingPanel title={ __( 'Hidden Lesson', 'wporg-learn' ) }>
+		<PluginDocumentSettingPanel title={ __( 'Hidden Lesson (deprecated)', 'wporg-learn' ) }>
```
Comment (adamwoodnz) [inline wp-content/plugins/wporg-learn/js/lesson-archive-excluded-meta/index.js:19]: Delete this setting panel block once data migration is complete

File: wp-content/themes/pub/wporg-learn-2024/inc/query.php [hunk 2]
```diff
@@ -52,8 +52,9 @@ function add_excluded_to_lesson_archive_query( $query ) {
 	// Ensure this code runs only for the main query on lesson archive pages and search results.
 	if ( ! is_admin() && $query->is_main_query() && ( $query->is_archive( 'lesson' ) || $query->is_search() ) ) {
 		$meta_query = $query->get( 'meta_query', array() );
+		$tax_query = $query->get( 'tax_query', array() );
 
-		$exclude_lesson_query = array(
+		$exclude_lessons_by_post_meta = array(
```
Comment (adamwoodnz) [inline wp-content/themes/pub/wporg-learn-2024/inc/query.php:57]: Meta query can be removed once data migration is complete

**Comment by adamwoodnz** [comment]
@gibrown @kangzj this has shipped now, but doesn't seem to have fixed this issue with hidden lessons appearing in search results. This production lesson 'Quiz time' has the 'hidden' term from the existing Jetpack 'show' taxonomy applied. It is hidden when viewing the archive, but [visible in search](https://learn.wordpress.org/?s=quiz&post_type=lesson):



I've used the [jptools](https://jptools.wordpress.com/debug/?url=https://learn.wordpress.org/#async_section_search) to check the post (id 257231) and I can see that it has the taxonomy applied.

I have tried to do an [incremental sync](https://jptools.wordpress.com/debug/?url=https://learn.wordpress.org/#async_get_incremental_sync_status), but discrepancies are still reported:

```
WP.com Cache has discrepancies as of last checksum on 2024-08-09 00:38:37


WP.com Cache Table Accuracy
posts: 100.00%
postmeta: 99.91%
comments: 100.00%
commentmeta: 100.00%
terms: 100.00%
term_relationships: 98.09%
term_taxonomy: 100.00%
users: 99.41%
usermeta: 99.42%
```

Should I try a full sync?

I have tried this in the test site too with a full sync and [it's not working there either](https://learn.wordpress.org/test/?s=hidden&post_type=lesson), but does work with Jetpack Search disabled:

| JP search off | JP search on |
|--------|--------|
| Screenshot 2024-08-09 at 12 29 32 PM | Screenshot 2024-08-09 at 12 29 52 PM | 

Any idea what's going wrong?

Thanks in advance for your help.

**Comment by trakos** [comment]
I've replied in https://github.com/WordPress/Learn/issues/2735

**Pull request merged**